### ISSUE TITLE
False positive for *used-before-assignment checkers

### ISSUE THREAD

**Issue by skirpichev** [pylint-dev/pylint#9663]
### Bug description

It seems, these checkers could understand very simple logic.  Maybe they shouldn't be enabled per default?  

An example:
```python
# pylint: disable=invalid-name, missing-module-docstring
a = bool(input())
d = False

if a:
    b = 1
    c = 2
    d = True

if d:
    print(b + c)

if a:
    b -= c
```

In case the reproducer looks artificial, here is similar real-world scenario:
https://github.com/diofant/diofant/blob/2def774bbcf8bce75362404b867497b255e96c03/diofant/concrete/gosper.py#L152-L170

### Command used

```shell
pylint a.py
```

### Pylint output

```python
************* Module a
a.py:11:10: E0601: Using variable 'b' before assignment (used-before-assignment)
a.py:11:14: E0606: Possibly using variable 'c' before assignment (possibly-used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

No errors.

### Pylint version

```shell
pylint 3.2.2
astroid 3.2.2
Python 3.11.3+ (heads/3.11:9fbb614c4e, Apr 29 2023, 14:18:05) [GCC 10.2.1 20210110]
```

**Comment by jacobtylerwalls** [comment]
Thanks for the message. You may find that this check is not a good fit for your team. As ever, pylint isn't smarter than you.

Duplicate of #9628, where you will find more discussion of why in some simple examples that are guaranteed to never fail, this message is less of an error than a code style warning.
